pkg/controllers/inventory/inventory_controller.go

```diff
@@ -0,0 +1,210 @@
+package inventory
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/config"
+	"github.com/vmware-tanzu/nsx-operator/pkg/logger"
+	commonservice "github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/inventory"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+
+	// How long to wait before retrying
+	minRetryDelay = 5 * time.Second
+	maxRetryDelay = 300 * time.Second
+
+	inventoryGCJitterFactor = 0.1
+)
+
+var (
+	log     = &logger.Log
+	keyFunc = cache.DeletionHandlingMetaNamespaceKeyFunc
+)
+
+type InventoryQuerier interface {
+	GetNodeNameById(externalId string) string
+}
+
+type InventoryController struct {
+	Client               client.Client
+	service              *inventory.InventoryService
+	inventoryObjectQueue workqueue.TypedRateLimitingInterface[any]
+	keyBuffer            inventory.KeySet
```

Comment:
I think keyBuffer is redundant, inventoryObjectQueue is enough, if you use channel, you don't need to get obj from key again at all, The pseudocode is as follows

func (b *BatchProcessor) Run(stopCh <-chan struct{}) {
    batch := make([]NSXOperation, 0, batchSize)
    ticker := time.NewTicker(flushInterval)
    
    for {
        select {
        case op := <-b.operations:
            batch = append(batch, op)
            if len(batch) >= batchSize {
                b.processBatch(batch)
                batch = batch[:0]
            }
        case <-ticker.C:
            if len(batch) > 0 {
                b.processBatch(batch)
                batch = batch[:0]
            }
        case <-stopCh:
            return
        }
    }
}

Reply:
it's more clear to put add/delete in different buffer. when handling the request, it's not needed to check the operation again